File: src/main/java/iteration1/StudentSemester.java
package iteration1;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;

@JsonIgnoreProperties(value={"letterGrades","courses"},allowGetters = true)
@JsonPropertyOrder({"semesterNo","yano","takenCredit","completedCredit","note","givenCredit"})
public class StudentSemester {
    @JsonProperty("semesterNo")
    private int semesterNo;
    @JsonProperty("yano")
    private float yano;
    @JsonProperty("takenCredit")
    private int takenCredit;
    @JsonProperty("completedCredit")
    private int completedCredit;
    @JsonProperty("note")
    private float note;
    @JsonProperty("givenCourses")
    ArrayList<GivenCourse> givenCourses;
    List<String> letterGrades;

    public StudentSemester(int semesterNo) {
        letterGrades = new ArrayList<>();
        this.semesterNo = semesterNo;
    }

    public StudentSemester(){

    }

    public void calculateYano() {
        //To calculate yano, we have to know total credit of that semester.
        calculateCredit();

        this.note=0;
        for(GivenCourse course : givenCourses){
           this.note += course.getCredit() * course.getGrade();
        }
        this.yano = this.note/completedCredit;
    }

    public void calculateCredit() {
        completedCredit = 0;
        for(GivenCourse course : givenCourses){
            completedCredit += course.getCredit();
        }
    }

    public void calculateLetterGrade() {
        for(GivenCourse course : givenCourses){
            Float aFloat = course.getGrade();
            if (aFloat == 0) {
                letterGrades.add("FF");
            } else if (aFloat == 0.5) {
                letterGrades.add("FD");
            } else if (aFloat == 1) {
                letterGrades.add("DD");
            } else if (aFloat == 1.5) {
                letterGrades.add("DC");
            } else if (aFloat == 2) {
                letterGrades.add("CC");
            } else if (aFloat == 2.5) {
                letterGrades.add("CB");
            } else if (aFloat == 3) {
                letterGrades.add("BB");
            } else if (aFloat == 3.5) {
                letterGrades.add("BA");
            } else if (aFloat == 4) {
                letterGrades.add("AA");
            }
        }

    }
    public void setYano(float yano) {
        this.yano = yano;
    }

    public float getYano() {
        return yano;
    }

    public int getTakenCredit() {
        return takenCredit;
    }

    public void setTakenCredit(int takenCredit) {
        this.takenCredit = takenCredit;
    }

    public int getCompletedCredit() {
        return completedCredit;
    }

    public void setCompletedCredit(int completedCredit) {
        this.completedCredit = completedCredit;
    }

    public void setNote(float note) {
        this.note = note;
    }

    public float getNote() {
        return note;
    }

    public int getSemesterNo() {
        return semesterNo;
    }

    public void setSemesterNo(int semesterNo) {
        this.semesterNo = semesterNo;
    }

    public ArrayList<GivenCourse> getGivenCourses() {
        return givenCourses;
    }

    public void setGivenCourses(ArrayList<GivenCourse> givenCourses) {
        this.givenCourses = givenCourses;
    }

    public List<String> getLetterGrades() {
        return letterGrades;
    }

    public void setLetterGrades(List<String> letterGrades) {
        this.letterGrades = letterGrades;
    }

    @Override
    public String toString() {
        return "StudentSemester{" +
                "semesterNo=" + semesterNo +
                ", yano=" + yano +
                ", completedCredit=" + completedCredit +
                ", note=" + note +
                ", givenCourses=" + givenCourses +
                ", letterGrades=" + letterGrades +
                '}';
    }
}

File: StudentSemester.py
class StudentSemester:
    def __init__(self, semester_no:int):
        self.semester_no = semester_no
        self.yano = None
        self.taken_credit = 0
        self.completed_credit = 0
        self.note = 0
        self.given_courses = []
        self.letter_grades = []

    def calculate_yano(self):
        self.calculate_cumulative_credit()
        self.note = 0
        for course in self.given_courses:
            self.note += course.credit * course.grade
        if self.completed_credit == 0:
            self.yano = 0
        else:
            self.yano = int((self.note / self.completed_credit) * 100) / 100.0

    def calculate_cumulative_credit(self):
        self.completed_credit = 0
        for course in self.given_courses:
            self.completed_credit += course.credit

    def calculate_letter_grade(self):
        for course in self.given_courses:
            course.calculate_letter_grade()
            if course.grade == 0:
                self.letter_grades.append("FF")
            elif course.grade == 0.5:
                self.letter_grades.append("FD")
            elif course.grade == 1:
                self.letter_grades.append("DD")
            elif course.grade == 1.5:
                self.letter_grades.append("DC")
            elif course.grade == 2:
                self.letter_grades.append("CC")
            elif course.grade == 2.5:
                self.letter_grades.append("CB")
            elif course.grade == 3:
                self.letter_grades.append("BB")
            elif course.grade == 3.5:
                self.letter_grades.append("BA")
            elif course.grade == 4:
                self.letter_grades.append("AA")

@property
def yano(self):
    return self.yano

@yano.setter
def yano(self, yano):
    self.yano = yano


@property
def taken_credit(self):
    return self.taken_credit


@taken_credit.setter
def taken_credit(self, taken_credit):
    self.taken_credit = taken_credit


@property
def completed_credit(self):
    return self.completed_credit


@completed_credit.setter
def completed_credit(self, completed_credit):
    self.completed_credit = completed_credit


@property
def note(self):
    return self.note


@note.setter
def note(self, note):
    self.note = note


@property
def semester_no(self):
    return self.semester_no


@semester_no.setter
def semester_no(self, semester_no):
    self.semester_no = semester_no


@property
def given_courses(self):
    return self.given_courses


@given_courses.setter
def given_courses(self, given_courses):
    self.given_courses = given_courses


@property
def letter_grades(self):
    return self.letter_grades


@letter_grades.setter
def letter_grades(self, letter_grades):
    self.letter_grades = letter_grades


def __str__(self):
    return "StudentSemester{" + \
           "semesterNo=" + str(self.semester_no) + \
           ", yano=" + str(self.yano) + \
           ", takenCredit=" + str(self.taken_credit) + \
           ", completedCredit=" + str(self.completed_credit) + \
           ", note=" + str(self.note) + \
           ", givenCourses=" + str(self.given_courses) + \
           ", letterGrades=" + str(self.letter_grades) + \
           '}'
